Generate code to check pseudo-class/element argument

nt1m opened the pull request:
#### e0c29e7e482a4a452c92c394677358659c71eace
<pre>
Generate code to check pseudo-class/element argument
<a href="https://bugs.webkit.org/show_bug.cgi?id=267029">https://bugs.webkit.org/show_bug.cgi?id=267029</a>
<a href="https://rdar.apple.com/120404097">rdar://120404097</a>

Reviewed by Ryosuke Niwa.

Add a field in `CSSPseudoSelectors.json` that represents whether a pseudo takes a required or optional argument, and use it to:
- generate `isOnlyPseudoClassFunction` as `CSSSelector::pseudoClassRequiresArgument`
- generate `isOnlyPseudoElementFunction` as `CSSSelector::pseudoElementRequiresArgument`
- add ASSERTs when serialization code hasn&apos;t been provided for pseudos with an optional or required argument

* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumeAttribute):
(WebCore::CSSSelectorParser::consumePseudo):
(WebCore::isOnlyPseudoClassFunction): Deleted.
(WebCore::isOnlyPseudoElementFunction): Deleted.
* Source/WebCore/css/process-css-pseudo-selectors.py:

Canonical link: <a href="https://commits.webkit.org/272601@main">https://commits.webkit.org/272601@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6766b478bd738af0713f87ad0cdd0f4d00ce3053

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32326 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11063 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34147 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34880 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/29265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33176 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13415 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8269 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32744 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9329 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28889 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8122 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8262 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28820 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36221 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29390 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29249 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/34388 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8391 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6337 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/32251 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10050 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/9022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4179 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8955 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/css/CSSSelector.h:
```diff
@@ -118,7 +118,11 @@ class CSSSelector {
+    static bool pseudoClassRequiresArgument(PseudoClass);
+    static bool pseudoElementRequiresArgument(PseudoElement);
 
+    static bool hasCustomSelectorText(PseudoClass);
```
This function name is rather vague. How about something like `pseudoClassMayHaveArgument`?

nt1m replied:
Good idea, renamed!

nt1m committed:
Generate code to check pseudo-class/element argument
https://bugs.webkit.org/show_bug.cgi?id=267029
rdar://120404097

Reviewed by Ryosuke Niwa.

Add a field in `CSSPseudoSelectors.json` that represents whether a pseudo takes a required or optional argument, and use it to:
- generate `isOnlyPseudoClassFunction` as `CSSSelector::pseudoClassRequiresArgument`
- generate `isOnlyPseudoElementFunction` as `CSSSelector::pseudoElementRequiresArgument`
- add ASSERTs when serialization code hasn't been provided for pseudos with an optional or required argument

* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumeAttribute):
(WebCore::CSSSelectorParser::consumePseudo):
(WebCore::isOnlyPseudoClassFunction): Deleted.
(WebCore::isOnlyPseudoElementFunction): Deleted.
* Source/WebCore/css/process-css-pseudo-selectors.py:

Canonical link: https://commits.webkit.org/272601@main
